### DISCUSSION TITLE
mdast-util-to-markdown: expected output with nested lists

### DISCUSSION THREAD

**Discussion by triaeiou** [syntax-tree/unist#77] [Q&A] [answered]
Hi (I seem to be asking a lot of questions lately),

The following
```
import {fromHtml as hastFromHtml} from 'hast-util-from-html';
import {toMarkdown as mdastToMarkdown} from 'mdast-util-to-markdown';
import {toMdast as hastToMdast} from 'hast-util-to-mdast';

const html = '<ul><li>one</li><li>two</li></ul>';

const hast = hastFromHtml(html);
const mdast = hastToMdast(hast);
const md = mdastToMarkdown(mdast);
console.log(md);
```
results in (i.e. no blank rows between list items)
```
*   one
*   two
```

Is the following the expected output from `const html = '<ul><li>one</li><li>two<ul><li>1</li><li>2</li></ul></li></ul>' ;` (i.e. a blank row between root level list items)?
```
*   one

*   two

    *   nested one
    *   nested two
```

Thanks

**Comment by wooorm** [comment, marked as answer]
Hey. This is vague. If you want to improve the way lists are made loose or tight, which might be possible, a proposal on how it would work would be more useful?

**Reply by triaeiou** [reply]
I thought I'd confirm that this might not be the expected behavior before I looked into possible mitigations (setting `tightDefinitions: true` does not seem to result in any change in this case).

My preferred output without tightDefinitions would be
```
*   one
*   two

    *   nested one
    *   nested two
``` 
and with tightDefinitions 
```
*   one
*   two
    *   nested one
    *   nested two
``` 

Thoughts?

**Reply by wooorm** [reply]
> does not seem to result in any change in this case

`tightDefinitions` is about definitions. Not lists.

Definitions are this:

```
a [link][ref].

[ref]: https://example.com
```

> My preferred output without tightDefinitions would be

You are talking about loose and tight lists.
Loose and tight have to do with how they are authored.
In HTML they result in whether `<p>` and `</p>` are generated or not.

See: https://spec.commonmark.org/0.30/#loose.
See: https://github.com/syntax-tree/mdast#list

As “blank line or not” is not a thing in HTML, it is a choice when generating.
This is done here: https://github.com/syntax-tree/hast-util-to-mdast/blob/bd47018ad2f516329b7495a3a499a8e212ee9f93/lib/util/list-items-spread.js#L9
And here: https://github.com/syntax-tree/hast-util-to-mdast/blob/bd47018ad2f516329b7495a3a499a8e212ee9f93/lib/handlers/li.js#L50.

**Reply by triaeiou** [reply]
> You are talking about loose and tight lists.

My bad.

> As “blank line or not” is not a thing in HTML, it is a choice when generating.

The commonmark dingus for
```
- one
- two
    - subone
    - subtwo
```
yields
```
<ul>
<li>one</li>
<li>two
<ul>
<li>subone</li>
<li>subtwo</li>
</ul>
</li>
</ul>
```
and
```
- one

- two
    - subone
    - subtwo
```
yields
```
<ul>
<li>
<p>one</p>
</li>
<li>
<p>two</p>
<ul>
<li>subone</li>
<li>subtwo</li>
</ul>
</li>
</ul>
```
To me then, the absence of `<p>` seems like it should result in tight lists, at least between list items at the same level. If you think that this is a reasonable I could look at how to achive that.

**Reply by wooorm** [reply]
I generally agree, and I am open to improving the situation!

Although it is a bit different: blank lines between or inside items, are not the same as authoring `<p>` or not, as the former can be made if there are no paragraphs too. So what would be done if there are, say, headings or code blocks in the items?

And, a question remains: how to detect if `<p>` tags were written?

**Reply by triaeiou** [reply]
Ok, looking at commonmark 301-326 the key for making a list loose is blank lines which tend to result in `<p>`-tags in the HTML output (excluding the "a list is loose when an item contains two block level elements separated by a blank line" - there is no HTML difference between blank line or not between two block elements). When converting the hast node to mdast the `<li>` children are wrapped in a mdast `paragraph` even if there was no `<p>` on the hast side to begin with (a design choice I suppose?). But checking prior to the conversion seems to work fine, something like this in [li.js](https://github.com/syntax-tree/hast-util-to-mdast/blob/bd47018ad2f516329b7495a3a499a8e212ee9f93/lib/handlers/li.js#L48.)

```
  const nd = clone || node;
  const spread = nd.children.some((el) => {
    return el.tagName === 'p';
  });
  const content = wrapChildren(h, nd);
  return h(node, 'listItem', {spread: spread, checked}, content)
```

The above seems to produce correct output with regards to loose/tight from commonmark 301-326 (i.e. convert the commonmark item HTML to markdown and compare that to the commonmark item markdown, the output is not identical but that is because the different variants in markdown that produce the same output in HTML).

There is one mismatch though (318 and 324), which seems to be unrelated to this, code blocks should be started on new lines but in the output are started on the same line as the list item marker, i.e. `<ul><li><pre><code>some_code();</code></pre></li><li>regular item</li></ul>` should produce

```
*
····somecode();
* regular item
```

but instead produces

```
*·····somecode();
* regular item
```

Thoughts?

**Reply by wooorm** [reply]
> Thoughts

That difference in output does not matter, I believe?
CommonMark is testing weird output here. `mdast-util-to-markdown` is generating pretty output?

> a design choice I suppose?

Yes, it is. To simplify (handling) the AST.
Similarly, in HTML, whether `<p>` and `</p>` exist does not matter whether “paragraphs” as a concept form. You can omit the tags in these cases and to HTML the text is still a paragraph content. The difference is for CSS and scripting: there is no specific element for it.\*

---

There is also a design decision in mdast where `spread` is used, on *both* lists and list items.
That way we can better match how markdown was written (or is supposed to be generated), rather than what happens when generating HTML.

---

What should we do if one list item contains a heading and a code block, so no paragraphs? Should it spread? What if next to it is an item with a paragraph? Or with multiple?

**Reply by triaeiou** [reply]
>That difference in output does not matter, I believe?

I assume you are referencing the non-newlined code blocks? From my use case it doesn't matter (because there is no code block available per default in the editors) but for completeness I think it should newline - however it seems unrelated to tight/loose lists.

> CommonMark is testing weird output here. mdast-util-to-markdown is generating pretty output?

Attached is a [print out](https://github.com/syntax-tree/unist/files/9310655/l.txt) - it's not a 100% test, I have taken the commonmark [output HTML](https://github.com/syntax-tree/unist/files/9310741/cm_list.txt) as source for `hast-util-from-html → hast-util-to-mdast → mdast-util-to-markdown` , but it gives some indication.

> You can omit the tags in these cases and to HTML the text is still a paragraph content. The difference is for CSS and scripting: there is no specific element for it.*

Well, there is a difference in the DOM as well as in how it is rendered without any CSS but that is fine.

> What should we do if one list item contains a heading and a code block, so no paragraphs? Should it spread? What if next to it is an item with a paragraph? Or with multiple?

The following is of no importance to my personal use case but for completeness I suggest following commonmarks idea which I interpret as a list is tight unless:
- There is a blank line prior to the next item ⇒ the item has a `<p>` as last child (which may also be the only child), one could argue that a trailing `<br>` should also qualify. I.e. "a heading and a code block, no paragraphs" = no spread regardless of neighbors (however the heading needs to be output correctly etc.)
- It has two direct children that are block elements that are separated by a blank line ⇒ `!phrasing()` - `<p>` - `!phrasing()`, possibly the `<p>` could also be a `<br>`

Right now the above logic is simplified to "contains a `<p>`" but that can be amended.

**Reply by wooorm** [reply]
> but for completeness I think it should newline - however it seems unrelated to tight/loose lists.

This seems like a separate discussion, which we can have, but we don‘t need to ;)
In short: mdast is an AST, so it’s going to loose info. It can’t roundtrip completely, and that’s fine.
That blank line is one such “meaningless“ thing, which it doesn’t roundtrip. And I think that’s fine.

> Attached is a [print out](https://github.com/syntax-tree/unist/files/9310655/l.txt)

P.S., I like using Gists for attachments like this!

---

301 looks weird? How is `mdast-util-to-markdown` turning an unordered list into an ordered one?
302: Edge case, that can be solved with https://github.com/syntax-tree/mdast-util-to-markdown#optionsbulletorderedother. I like always using one marker, and forcing a break with an empty comment tho.
303: More clear.
304: Is `mdast-util-to-markdown` removing the EOL? 🤔
305/306/307: I think the output from `mdast-util-to-markdown` is an improvement.
308: Again an unordered list turning into an ordered one?
...rest I often don’t understand, or IMO better, too?

> Well, there is a difference in the DOM as well as in how it is rendered without any CSS but that is fine.

Right. Not to semantics, but a different tree, that you can select with CSS (or JS I guess).

> but for completeness

Yeah, I’m looking for a good solution that works for everyone, outside of your specific use case! A real improvement, rather than just a change.

> There is a blank line prior to the next item 

If I understand what you are proposing correctly, that would not work (well) on minified HTML. I think it is very important to work on minified HTML too.

---

As an alternative, what if we detect that phrasing content only, is used directly inside a `<li>`? Similar to your recent patch, use `hast-util-phrasing` to check that (and some extra handling to ignore comments, the doctype)?

**Reply by triaeiou** [reply]
> 301 looks weird? How is mdast-util-to-markdown turning an unordered list into an ordered one?
> 308: Again an unordered list turning into an ordered one?

Sorry that's due to my unclear log print out (I regexed out the commonmark samples into a dictionary which i looped through and printed the commonmark markdown and the `mdast-to-markdown`  output for each - however as mentioned I sometimes changed the HTML, without changing the commonmark markdown, to make the markdown print out more clearer in list separations) in 301 and 308 I changed `<ul>` to `<ol>` to see that the list separation was where it should be.

> 304: Is mdast-util-to-markdown removing the EOL? 🤔

That's a soft break I believe so it doesn't have any `<br>` in the HTML and so disappears in HTML ⇒ markdown.

> 305/306/307: I think the output from mdast-util-to-markdown is an improvement.

I think those commonmark samples are to demonstrate syntax latitude, not "standard" and `mdast-util-to-markdown` obviously outputs "standard"

>> There is a blank line prior to the next item

> If I understand what you are proposing correctly, that would not work (well) on minified HTML. I think it is very important to work on minified HTML too.

I was discussing the commonmark spec notes (regarding markdown syntax). To translate that to HTML I was suggesting that `<ul><li>Some text<br></li></ul>`, `<ul><li><p>Some text</p></li></ul>` and `<ul><li><ul><li>Nested text</li></ul><p>Non-nested text</p></li></ul>` would result in loose lists (the outer list in case 3, inner is tight). That would work the same on any format HTML I believe?


> As an alternative, what if we detect that phrasing content only, is used directly inside a `<li>`? Similar to your recent patch, use hast-util-phrasing to check that (and some extra handling to ignore comments, the doctype)?

I think that wouldn't meet these differences?
```
- # A header
  with some text
- and next item
```
outputs `<ul><li><h1>A header</h1>with some text</li><li>and next item</li></ul>` (newlines removed) in the commonmark dingus whereas
```
- # A header
  with some text

- and next item
```
outputs `<ul><li><h1>A header</h1><p>with some text</p></li><li><p>and next item</p></li></ul>`

Granted, I think these are minor details. It would be "neat" to be able to run the commonmark samples in both directions but the question is how useful that is for  HTML ⇒ markdown as there are so many more variants of HTML that would produce similar looking output.

**Reply by wooorm** [reply]
Hmm, I don‘t really understand exactly what you are proposing here now. But perhaps a PR to `hast-util-to-mdast` will help clear that confusion out. And then we have a lot of tests to see if it breaks or improves things.

For the `<li>Some text<br></li></ul>` example, according to HTML, a `<br>` should not be used like that: it should always be preceded and followed by text. That’s why it can’t occur in markdown. And why they are [filtered out by us](https://github.com/syntax-tree/hast-util-to-mdast/commit/a02be0ec75bf66f01e3035add243e942ff67adfe). So I don’t think we should worry or care about them!

> I think that wouldn't meet these differences?

> Granted, I think these are minor details. It would be "neat" to be able to run the commonmark samples in both directions but the question is how useful that is for HTML ⇒ markdown as there are so many more variants of HTML that would produce similar looking output.

Correct. I don’t think that is very important. I don’t think we can roundtrip all CommonMark output examples to their input anyway? The goal here I believe is to create good and readable markdown that resembles the intent of the HTML.

**Reply by triaeiou** [reply]
Ok, I've thought about it some more and made a "demo" [PR](https://github.com/syntax-tree/hast-util-to-mdast/pull/75) with minimal changes (for better code clarity one should probably rename one variable but in this branch I kept the edit to as few lines as possible, besides `npm test` fails as it compares to remark output which I guess has the same logic).

The [current](https://github.com/syntax-tree/hast-util-to-mdast/blob/fa388c75467d8633ebe0b7c0abfa8638bccdc70b/lib/handlers/li.js#L50) logic is: *a list is loose if at least one list item contains more than one child*. This interprets `<ul><li>Outer 1<ul><li>Inner 1</li><li>Inner 2</li></ul></li><li>Outer 2</li></ul>` outer list as loose even though it is the output from a tight list (had it been the output of a loose outer list `Outer 1` would have been wrapped in a `<p>`.

My suggested logic, when disregarding any `<br>`,  is *a list is loose if at least one list item contains a `<p>` (which is the output effect of a markdown loose list). Granted this logic misses lists that were loose in Markdown but contained no paragraphs, e.g.
```
- # Outer 1

  # Second block
- # Outer 2
```
The spec says *A list is loose [snip] any of its constituent list items directly contain two block-level elements with a blank line between them* but two block level elements with a blank line becomes the same HTML as without.

**Reply by wooorm** [reply]
Response in the PR: https://github.com/syntax-tree/hast-util-to-mdast/pull/75#issuecomment-1242676768.

**Comment by JounQin** [comment]
````md
- one
- two
    - subone
    - subtwo

---

- one

- two
    - subone
    - subtwo
````

Rendered:

- one
- two
    - subone
    - subtwo

---

- one

- two
    - subone
    - subtwo

---

They're just working as expected here AFAIK.

**Reply by triaeiou** [reply]
> They're just working as expected here AFAIK.

How do you mean? Just to avoid confusion, the dingus reference was just to illustrate what the reverse does, the discussion is about what markdown to output when coming from HTML/mdast.

**Reply by JounQin** [reply]
See the source codes and rendered result on GitHub at https://github.com/orgs/syntax-tree/discussions/77#discussioncomment-3368384

**Reply by triaeiou** [reply]
> See the source codes and rendered result on GitHub at https://github.com/orgs/syntax-tree/discussions/77#discussioncomment-3368384

Yes, but the issue is going the other direction, i.e. what markdown to _output_ from `<ul><li>one</li><li>two<ul><li>subone</li><li>subtwo</li></ul></li></ul>`

What "seems" to be the reverse behaviour of commonmark and micromark:
```
- one
- two

    - subone
    - subtwo
```
or the current implementation of `toMarkdown()`:
```
- one

- two

    - subone
    - subtwo
```
